monads/src/HW/Eval.hs

```diff
@@ -4,33 +4,66 @@ import qualified Data.Map as M
 import State.MyState
 import HW.StackMachine
 
--- Possible errors during evaluation
 data Error v
-  = StackUnderflow (StackInstr v) -- Not enough numbers on the stack to process the instruction
-  | VarUndefined String           -- The variable is not defined in the environment 
-  | StackNotExhausted Stack       -- After the program has finished evaluation, the stack is not a singleton
+  = StackUnderflow (StackInstr v)
+  | VarUndefined String
+  | StackNotExhausted Stack
+  | FinalStackEmpty
   deriving (Show, Eq)
 
-type Stack = [Int] -- The stack, holding integers
 
-type Env v = M.Map v Int -- The environment, mapping variables to their values
+type Stack = [Int]
+
+type Env v = M.Map v Int
 
--- The machine state consists of the stack and the variable environment.
 data MachineState v = MachineState
   { getStack :: Stack,
     getEnv :: Env v
   }
-  deriving (Show)
+  deriving (Show, Eq)
 
--- Run the compiled program on an empty stack and environment
 initialState :: MachineState String
 initialState = MachineState [] M.empty
 
--- Execute a single instruction. 
--- Successful evaluation does not produce any useful result: only the effect of modifying state matters. 
 execInstr :: (Ord v, Show v) => StackInstr v -> MyState (MachineState v) (Either (Error v) ())
-execInstr = undefined 
+execInstr instr = do
+  MachineState stack env <- get
```

Comment:
It's better to use `modify` instead of `put`, and only get the subpart of the state you are going to use. This way is less error-prone.